qiime2/sdk/parallel_context.py

```diff
@@ -0,0 +1,247 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2023, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from parsl.app.app import python_app
+
+
+from qiime2.sdk.context import Context
+from qiime2.sdk.action import Pipeline
+from qiime2.sdk.proxy import Proxy, ProxyResults
+from qiime2.sdk.parallel_config import PARALLEL_CONFIG
+
+
+def _map_arg(arg, futures):
+    """ Map a proxy artifact for input to a parsl action
+    """
+    # We add this future to the list and create a new proxy with its index as
+    # its future.
+    if isinstance(arg, Proxy):
+        futures.append(arg._future_)
+        mapped = arg.__class__(len(futures) - 1, arg._selector_)
+    # We do the above but for all elements in the collection
+    elif isinstance(arg, list):
+        mapped = []
+
+        for proxy in arg:
```

Comment:
I know this is older code, but the guts of this for-loop and the dict can be replaced with a recursive comprehension:

```python
if isinstance(arg, Proxy):
    ...
elif isinstance(arg, list):
    mapped = [_map_arg(proxy, futures) for proxy in arg]
elif isinstance(arg, dict):
    mapped = {key: _map_arg(proxy, futures) for key, proxy in arg.items()}
```

Reply:
Good call